### DISCUSSION TITLE
What's my problem; I can't burn bootlooder or load programs?

### DISCUSSION THREAD

**Discussion by microPaul** [SpenceKonde/DxCore#217] [Q&A]
I'm never sure if it's me or the software.

I've updated DXCore to 1.4.1, tinyMegaCore to 2.5.2 and the Arduino IDE to 1.8.19.  I have my homebrew DB64 board that was working fine with Arduino IDE and DX core, but something must have changed because now I can't get programs into the device using the Arduino IDE.  I thought it might be my board, but I can use MPLAB X IDE and SNAP to get blink sketches into the board and run successfully, which makes me think the problem is with the IDE or the Core (or the python transfer program).

I don't understand why the Arduino IDE is squawking about JAVA but it is.  Is there some sort of a lock file it expects to find but can't, or maybe a lock file that should be unlocked is found to be locked?

Here's what I'm getting from the Arduino IDE....

`Sketch uses 4310 bytes (3%) of program storage space. Maximum is 131072 bytes.
Global variables use 401 bytes (2%) of dynamic memory, leaving 15983 bytes for local variables. Maximum is 16384 bytes.
java.io.IOException: Cannot run program "{upload.prog_interlock}\"C:\Users\p\AppData\Local\Arduino15\packages\megaTinyCore\tools\python3\3.7.2-post1/python3\"": CreateProcess error=2, The system cannot find the file specified
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at processing.app.helpers.ProcessUtils.exec(ProcessUtils.java:26)
	at cc.arduino.packages.Uploader.executeUploadCommand(Uploader.java:129)
	at cc.arduino.packages.uploaders.SerialUploader.runCommand(SerialUploader.java:383)
	at cc.arduino.packages.uploaders.SerialUploader.uploadUsingProgrammer(SerialUploader.java:315)
	at cc.arduino.packages.uploaders.SerialUploader.uploadUsingPreferences(SerialUploader.java:90)
	at cc.arduino.UploaderUtils.upload(UploaderUtils.java:77)
	at processing.app.SketchController.upload(SketchController.java:732)
	at processing.app.SketchController.exportApplet(SketchController.java:703)
	at processing.app.Editor$UploadHandler.run(Editor.java:2061)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: CreateProcess error=2, The system cannot find the file specified
	at java.lang.ProcessImpl.create(Native Method)
	at java.lang.ProcessImpl.<init>(ProcessImpl.java:386)
	at java.lang.ProcessImpl.start(ProcessImpl.java:137)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 10 more
An error occurred while uploading the sketch
`

Any suggestions would be greatly appreciated.

Paul

**Comment by ITDiscovery** [comment]
As a temporary fix, run the command after:
```
"{upload.prog_interlock}\
```

From the command line, which worked for me. I also recently upgraded and have the same problem. I suspect it's some sort of path/permissions issue, because I can indeed run from the command line. Here's what I am seeing:

```
{upload.prog_interlock}"/Users/xxxxxxxxx/Library/Arduino15/packages/megaTinyCore/tools/python3/3.7.2-post1/python3" -u /Users/xxxxxxxxxx/Library/Arduino15/packages/DxCore/hardware/megaavr/1.4.2/tools/prog.py -t uart -u /dev/cu.usbserial-AB0LX43L -b 230400 -d avr128db28 --fuses 5:0b11001001 6:0b00001100 7:0x00 8:0x00 -f/var/folders/lf/vbl7v6c55ll273ckcmn48gqw0000gp/T/arduino_build_178299/AVR128DBTinyBasic.ino.hex -a write 
java.io.IOException: Cannot run program "{upload.prog_interlock}"/Users/xxxxxxxxx/Library/Arduino15/packages/megaTinyCore/tools/python3/3.7.2-post1/python3"": error=2, No such file or directory
```

I pasted the command into terminal on my Mac...works fine. If your on Windows, and that command works fine...that should be a huge hint as to what's wrong. I'm on DxCore 1.4.2, the megaTinycore isn't installed on this box, but is on my other Mac and didn't make a difference. I'm on 1.8.20 for Arduino IDE. DxCore 1.3.8 (?) was working fine... Can someone try 1.4.2 with an older Arduino IDE?

My bad guess...it is missing a file, but not one in the error.

**Comment by SpenceKonde** [comment]
Argh.... 

I messed up porting a change from megaTinyCore - but that should absolutely not impact bootloading, only uploading to non-optiboot boards. - upload.prog_interlock isn't defined. It's supposed to be defined as an empty string for non-optiboot boards, but defined as an error message for optiboot boards, which will prevent you from doing "upload using programmer" which on modern AVRs, for boards set to use a bootloader, will result in the bootloader being removed... but the sketch uploaded will restart every millisecond, because the sketch will be located with it's vector table starting at 0x0200. but as far as the hardware is concerned, the vector table starts at 0x0000 if no bootloader is in use. I'll push out another emergency release tomorrow night. or during the day if the movers flake out on us again.

**Reply by SpenceKonde** [reply]
For future reference, this **definitely** is the kind of thing that you want to create an issue for. Discussions are a lot easier to miss, while I treat the issues list as the list of action items.

**Reply by microPaul** [reply]
Thank you Peter and Spence for this information.

Next time I will use Issue instead of discussion.  I was hesitant to pull the fire alarm (Issue) because I thought it may have been something I messed up on my PC, because there are some other things (not DX Core or megaTinyCore releated) that were not (and still are not) working correctly on my main PC.

**Reply by ITDiscovery** [reply]
Thanks for getting this corrected Spence!! Your post makes perfect sense, as I could indeed burn a boot loader without issues.

**Reply by ITDiscovery** [reply]
FYI still broke under 1.4.3. Off to grandma's house and will check back in a few days.

**Comment by SpenceKonde** [comment]
Yeah 1.4.3 doesn't have the fix yet, that corrects a different issue

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by SpenceKonde** [reply]
No shit it is reason for an emergency release, I just did three emergency bugfix releases over the prior few days. It just so happens that I found out that this problem was a bug in the core after the 1.4.3 emergency fix, and had to get to bed because the movers were getting here at 8am the next morning. I sent the above message during the pause while they finished loading the truck, and today is the next time that I've had something to put the computer ontop of.

**Reply by ITDiscovery** [reply]
Tested as fixed in 1.4.5 on my Mac. Many thanks Spence!!
